src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -15,20 +13,24 @@ internal sealed class ServiceProviderEngineScope : IServiceScope, IServiceProvid
         // For testing only
         internal Action<object> _captureDisposableCallback;
 
-        private List<object> _disposables;
-
         private bool _disposed;
-        private readonly object _disposelock = new object();
+        private ScopePool.State _state;
+
+        // This protects the disposed state
+        private readonly object _disposeLock = new object();
+
+        // This protects resolved services, this is only used if isRoot is false
+        private readonly object _scopeLock = new object();
 
         public ServiceProviderEngineScope(ServiceProviderEngine engine, bool isRoot = false)
         {
             Engine = engine;
-
-            // To reduce lock contention for singletons upon resolve we use a concurrent dictionary.
-            ResolvedServices = isRoot ? new ConcurrentDictionary<ServiceCacheKey, object>() : new Dictionary<ServiceCacheKey, object>();
+            _state = isRoot ? new ScopePool.State() : engine.ScopePool.Rent();
         }
 
-        internal IDictionary<ServiceCacheKey, object> ResolvedServices { get; }
+        internal IDictionary<ServiceCacheKey, object> ResolvedServices => _state?.ResolvedServices ?? ScopeDisposed();
+
+        internal object Sync => _scopeLock;
```

Comment:
Can we reuse the dispose lock?

Reply:
Yea I think we can do this now that singletons don't use this lock anymore.